File: backend/app/utils/base36.py
"""
Base-36 encoding/decoding utilities.
Uses standard base-36 character set: A-Z (0-25), 0-9 (26-35)
"""

# Base-36 character set: A-Z, 0-9
BASE36_CHARS = 'ABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789'


def encode_base36(number: int, min_length: int = 0) -> str:
    """
    Encode a number to base-36 string.
    
    Args:
        number: Integer to encode (must be >= 0)
        min_length: Minimum length of output string (padded with 'A' on the left)
    
    Returns:
        Base-36 encoded string
    
    Raises:
        ValueError: If number is negative
    """
    if number < 0:
        raise ValueError("Number must be non-negative")
    
    if number == 0:
        result = 'A'
    else:
        result = ''
        while number > 0:
            result = BASE36_CHARS[number % 36] + result
            number //= 36
    
    # Pad with 'A' to minimum length
    if len(result) < min_length:
        result = 'A' * (min_length - len(result)) + result
    
    return result


def decode_base36(string: str) -> int:
    """
    Decode a base-36 string to number.
    
    Args:
        string: Base-36 encoded string (case-insensitive)
    
    Returns:
        Decoded integer
    
    Raises:
        ValueError: If string contains invalid characters
    """
    string = string.upper()
    result = 0
    
    for char in string:
        if char not in BASE36_CHARS:
            raise ValueError(f"Invalid base-36 character: {char}")
        result = result * 36 + BASE36_CHARS.index(char)
    
    return result


def pad_base36(number: int, width: int) -> str:
    """
    Encode a number to base-36 and pad to specified width.
    
    Args:
        number: Integer to encode (must be >= 0)
        width: Desired width of output string (padded with 'A' on the left)
    
    Returns:
        Base-36 encoded string padded to width
    """
    return encode_base36(number, min_length=width)

File: frontend/pos/src/utils/base36.ts
/**
 * Base-36 encoding/decoding utilities.
 * Uses standard base-36 character set: A-Z (0-25), 0-9 (26-35)
 */

// Base-36 character set: A-Z, 0-9
const BASE36_CHARS = 'ABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789'

/**
 * Encode a number to base-36 string.
 * 
 * @param number - Integer to encode (must be >= 0)
 * @param minLength - Minimum length of output string (padded with 'A' on the left)
 * @returns Base-36 encoded string
 * @throws Error if number is negative
 */
export function encodeBase36(number: number, minLength: number = 0): string {
  if (number < 0) {
    throw new Error('Number must be non-negative')
  }
  let result = ''

  if (number === 0) {
    result = BASE36_CHARS[0]
  }
  else {
    let num = number
    while (num > 0) {
      result = BASE36_CHARS[num % 36] + result
      num = Math.floor(num / 36)
    }
  }

  // Pad with 'Zero' to minimum length
  if (result.length < minLength) {
    result = BASE36_CHARS[0].repeat(minLength - result.length) + result
  }

  return result
}

/**
 * Decode a base-36 string to number.
 * 
 * @param string - Base-36 encoded string (case-insensitive)
 * @returns Decoded integer
 * @throws Error if string contains invalid characters
 */
export function decodeBase36(string: string): number {
  const str = string.toUpperCase()
  let result = 0

  for (let i = 0; i < str.length; i++) {
    const char = str[i]
    const index = BASE36_CHARS.indexOf(char)
    if (index === -1) {
      throw new Error(`Invalid base-36 character: ${char}`)
    }
    result = result * 36 + index
  }

  return result
}

/**
 * Encode a number to base-36 and pad to specified width.
 * 
 * @param number - Integer to encode (must be >= 0)
 * @param width - Desired width of output string (padded with 'A' on the left)
 * @returns Base-36 encoded string padded to width
 */
export function padBase36(number: number, width: number): string {
  return encodeBase36(number, width)
}

